src/ape_ethereum/ecosystem.py

```diff
@@ -320,11 +324,113 @@ def create_transaction(self, **kwargs) -> TransactionAPI:
+                self.conversion_manager.convert(a, AddressType) if isinstance(a, AccountAPI) else a
+                for a in filter_args.values()
+            ]
+
+            # Add remaining kwargs as topics to filter on.
+            topics = LogInputs(abi, True)
+            encoded_topic_data = [
+                encode_single(topic_type, topic_data).hex()  # type: ignore
+                for topic_type, topic_data in zip(topics.types, search_topics)
+            ]
+            filter_data["topics"].extend(encoded_topic_data)
+        else:
+            filter_data["topics"] = filter_args.pop("topics")
+
+        return filter_data
+
+    def decode_logs(self, abi: EventABI, data: List[Dict]) -> Iterator["ContractLog"]:
+        if not abi.anonymous:
+            event_id_bytes = keccak(to_bytes(text=abi.selector))
+            matching_logs = [log for log in data if log["topics"][0] == event_id_bytes]
+            if not matching_logs:
+                raise DecodingError(f"No logs found with ID '{abi.selector}'.")
+
+        # Process indexed data (topics)
+        abi_topics = LogInputs(abi, indexed=True)
+        abi_data = LogInputs(abi, indexed=False)
+
+        # Verify no duplicate names
+        duplicate_names = set(abi_topics.names).intersection(abi_data.names)
+        if duplicate_names:
+            duplicate_names_str = ", ".join([n for n in duplicate_names if n])
+            raise DecodingError(
+                "The following argument names are duplicated "
+                f"between event inputs: '{duplicate_names_str}'."
+            )
+
+        for log in matching_logs:
+            indexed_data = log["topics"] if log.get("anonymous", False) else log["topics"][1:]
+            log_data = hexstr_if_str(to_bytes, log["data"])  # type: ignore
+
+            if len(indexed_data) != len(abi_topics.types):
+                raise DecodingError(
+                    f"Expected '{len(indexed_data)}' log topics.  Got '{len(abi_topics.types)}'."
+                )
+
+            decoded_topic_data = [
+                decode_single(topic_type, topic_data)  # type: ignore
+                for topic_type, topic_data in zip(abi_topics.types, indexed_data)
+            ]
+            decoded_log_data = decode_abi(abi_data.types, log_data)  # type: ignore
+            event_args = dict(
+                itertools.chain(
+                    zip(abi_topics.names, decoded_topic_data),
+                    zip(abi_data.names, decoded_log_data),
+                )
+            )
+            yield ContractLog(name=abi.name, data=event_args)  # type: ignore
+
+
+def _get_event_abi_types(abi_inputs: List[Dict]) -> Iterator[Union[str, Dict]]:
+    for abi_input in abi_inputs:
+        abi_type = grammar.parse(abi_input["type"])
+        if abi_type.is_dynamic:
+            yield "bytes32"
+        else:
+            yield collapse_if_tuple(abi_input)
+
+
+class LogInputs:
+    def __init__(self, abi: EventABI, indexed: bool):
```

Comment:
What does `indexed` mean here? I don't think this is the proper use of it, as each individual event arg may or may not be indexed. Is this supposed to be `anonymous`?

Reply:
Indexed means the index args! I am not sure I understand what you mean by this.
We need to discern the index from the non-index to properly decode.